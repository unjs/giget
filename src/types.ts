export interface GitInfo {
  provider: "github" | "gitlab" | "bitbucket" | "sourcehut" | "codeberg";
  repo: string;
  subdir: string;
  ref: string;
}

export interface TemplateInfo {
  name: string;
  tar: string;
  version?: string;
  subdir?: string;
  url?: string;
  defaultDir?: string;
  headers?: Record<string, string | undefined>;

  // Added by giget
  source?: never;
  dir?: never;

  [key: string]: any;
}

export type TemplateProvider = (
  input: string,
  options: { auth?: string },
) => TemplateInfo | Promise<TemplateInfo> | null;
