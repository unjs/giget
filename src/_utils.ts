import { createWriteStream, existsSync } from 'node:fs'
import { pipeline } from 'node:stream'
import { spawnSync } from 'node:child_process'
import { readFile, writeFile } from 'node:fs/promises'
import { promisify } from 'node:util'
import { relative, resolve } from 'pathe'
import { fetch } from 'node-fetch-native'
import createHttpsProxyAgent from 'https-proxy-agent'
import type { GitInfo } from './types'

export async function download (url: string, filePath: string, opts: { headers?: Record<string, string> } = {}) {
  const infoPath = filePath + '.json'
  const info: { etag?: string } = JSON.parse(await readFile(infoPath, 'utf8').catch(() => '{}'))
  const headRes = await sendFetch(url, { method: 'HEAD', headers: opts.headers }).catch(() => null)
  const etag = headRes?.headers.get('etag')
  if (info.etag === etag && existsSync(filePath)) {
    // Already downloaded
    return
  }
  info.etag = etag

  const res = await sendFetch(url, { headers: opts.headers })
  if (res.status >= 400) {
    throw new Error(`Failed to download ${url}: ${res.status} ${res.statusText}`)
  }

  const stream = createWriteStream(filePath)
  await promisify(pipeline)(res.body as any, stream)

  await writeFile(infoPath, JSON.stringify(info), 'utf8')
}

const inputRegex = /^(?<repo>[\w.-]+\/[\w.-]+)(?<subdir>[^#]+)?(?<ref>#[\w.-]+)?/

export function parseGitURI (input: string): GitInfo {
  const m = input.match(inputRegex)?.groups
  return <GitInfo> {
    repo: m.repo,
    subdir: m.subdir || '/',
    ref: m.ref ? m.ref.substring(1) : 'main'
  }
}

export function debug (...args) {
  if (process.env.DEBUG) {
    console.debug('[giget]', ...args)
  }
}

export async function sendFetch (url: string, options?: RequestInit) {
  const proxy = process.env.HTTPS_PROXY || process.env.https_proxy || process.env.HTTP_PROXY || process.env.http_proxy
  const requestOptions = proxy ? { agent: createHttpsProxyAgent(proxy), ...options } : options
  return await fetch(url, requestOptions)
}

// -- Experimental --

export function currentShell () {
  if (process.env.SHELL) {
    return process.env.SHELL
  }
  if (process.platform === 'win32') {
    return 'cmd.exe'
  }
  return '/bin/bash'
}

export function startShell (cwd: string) {
  cwd = resolve(cwd)
  const shell = currentShell()
  console.info(`(experimental) Opening shell in ${relative(process.cwd(), cwd)}...`)
  spawnSync(shell, [], {
    cwd,
    shell: true,
    stdio: 'inherit'
  })
}
