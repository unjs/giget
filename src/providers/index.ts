import { TemplateProvider } from "../types";
import http from "./http";
import github from "./github";
import gitlab from "./gitlab";
import bitbucket from "./bitbucket";
import sourcehut from "./sourcehut";

export const providers: Record<string, TemplateProvider> = {
  http,
  https: http,
  github,
  gh: github,
  gitlab,
  bitbucket,
  sourcehut,
};
