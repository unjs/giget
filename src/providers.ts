import { basename } from "pathe";
import type { TemplateInfo, TemplateProvider } from "./types";
import { debug, parseGitURI, sendFetch } from "./_utils";

export const http: TemplateProvider = async (input, options) => {
  if (input.endsWith(".json")) {
    return (await _httpJSON(input, options)) as TemplateInfo;
  }

  const url = new URL(input);
  let name: string = basename(url.pathname);

  try {
    const head = await sendFetch(url.href, {
      method: "HEAD",
      validateStatus: true,
      headers: {
        authorization: options.auth ? `Bearer ${options.auth}` : undefined,
      },
    });
    const _contentType = head.headers.get("content-type") || "";
    if (_contentType.includes("application/json")) {
      return (await _httpJSON(input, options)) as TemplateInfo;
    }
    const filename = head.headers
      .get("content-disposition")
      ?.match(/filename="?(.+)"?/)?.[1];
    if (filename) {
      name = filename.split(".")[0];
    }
  } catch (error) {
    debug(`Failed to fetch HEAD for ${url.href}:`, error);
  }

  return {
    name: `${name}-${url.href.slice(0, 8)}`,
    version: "",
    subdir: "",
    tar: url.href,
    defaultDir: name,
    headers: {
      Authorization: options.auth ? `Bearer ${options.auth}` : undefined,
    },
  };
};

const _httpJSON: TemplateProvider = async (input, options) => {
  const result = await sendFetch(input, {
    validateStatus: true,
    headers: {
      authorization: options.auth ? `Bearer ${options.auth}` : undefined,
    },
  });
  const info = (await result.json()) as TemplateInfo;
  if (!info.tar || !info.name) {
    throw new Error(
      `Invalid template info from ${input}. name or tar fields are missing!`,
    );
  }
  return info;
};

export const github: TemplateProvider = (input, options) => {
  const parsed = parseGitURI(input);

  // https://docs.github.com/en/rest/repos/contents#download-a-repository-archive-tar
  // TODO: Verify solution for github enterprise
  const githubAPIURL = process.env.GIGET_GITHUB_URL || "https://api.github.com";

  return {
    name: parsed.repo.replace("/", "-"),
    version: parsed.ref,
    subdir: parsed.subdir,
    headers: {
      Authorization: options.auth ? `Bearer ${options.auth}` : undefined,
      Accept: "application/vnd.github+json",
      "X-GitHub-Api-Version": "2022-11-28",
    },
    url: `${githubAPIURL.replace("api.github.com", "github.com")}/${
      parsed.repo
    }/tree/${parsed.ref}${parsed.subdir}`,
    tar: `${githubAPIURL}/repos/${parsed.repo}/tarball/${parsed.ref}`,
  };
};

export const gitlab: TemplateProvider = (input, options) => {
  const parsed = parseGitURI(input);
  const gitlab = process.env.GIGET_GITLAB_URL || "https://gitlab.com";
  return {
    name: parsed.repo.replace("/", "-"),
    version: parsed.ref,
    subdir: parsed.subdir,
    headers: {
      authorization: options.auth ? `Bearer ${options.auth}` : undefined,
      // https://gitlab.com/gitlab-org/gitlab/-/commit/50c11f278d18fe1f3fb12eb595067216bb58ade2
      "sec-fetch-mode": "same-origin",
    },
    url: `${gitlab}/${parsed.repo}/tree/${parsed.ref}${parsed.subdir}`,
    tar: `${gitlab}/${parsed.repo}/-/archive/${parsed.ref}.tar.gz`,
  };
};

export const bitbucket: TemplateProvider = (input, options) => {
  const parsed = parseGitURI(input);
  const bitbucketURL =
    process.env.GIGET_BITBUCKET_URL || "https://bitbucket.org";

  let paths: Pick<TemplateInfo, "url" | "tar"> = {
    url: `${bitbucketURL}/${parsed.repo}/src/${parsed.ref}${parsed.subdir}`,
    tar: `${bitbucketURL}/${parsed.repo}/get/${parsed.ref}.tar.gz`,
  };

  // Self-hosted Bitbucket paths
  if (bitbucketURL !== "https://bitbucket.org") {
    const [project, repo] = parsed.repo.split("/");

    paths = {
      url: `${bitbucketURL}/projects/${project}/repos/${repo}/browse${parsed.subdir}?at=${parsed.ref}`,
      tar: `${bitbucketURL}/rest/api/latest/projects/${project}/repos/${repo}/archive?at=${parsed.ref}&format=tar.gz`,
    };
  }

  return {
    name: parsed.repo.replace("/", "-"),
    version: parsed.ref,
    subdir: parsed.subdir,
    headers: {
      authorization: options.auth ? `Bearer ${options.auth}` : undefined,
    },
    ...paths,
  };
};

export const sourcehut: TemplateProvider = (input, options) => {
  const parsed = parseGitURI(input);
  return {
    name: parsed.repo.replace("/", "-"),
    version: parsed.ref,
    subdir: parsed.subdir,
    headers: {
      authorization: options.auth ? `Bearer ${options.auth}` : undefined,
    },
    url: `https://git.sr.ht/~${parsed.repo}/tree/${parsed.ref}/item${parsed.subdir}`,
    tar: `https://git.sr.ht/~${parsed.repo}/archive/${parsed.ref}.tar.gz`,
  };
};

export const providers: Record<string, TemplateProvider> = {
  http,
  https: http,
  github,
  gh: github,
  gitlab,
  bitbucket,
  sourcehut,
};
