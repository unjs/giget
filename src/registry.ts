import type { TemplateInfo, TemplateProvider } from './types'
import { debug, sendFetch } from './_utils'

// const DEFAULT_REGISTRY = 'https://cdn.jsdelivr.net/gh/unjs/giget/templates'
const DEFAULT_REGISTRY = 'https://raw.githubusercontent.com/unjs/giget/main/templates'

export const registryProvider = (registryEndpoint: string = DEFAULT_REGISTRY) => {
  return <TemplateProvider>(async (input) => {
    const start = Date.now()
    const registryURL = `${registryEndpoint}/${input}.json`

    const res = await sendFetch(registryURL)
    if (res.status >= 400) {
      throw new Error(`Failed to download ${input} template info from ${registryURL}: ${res.status} ${res.statusText}`)
    }
    const info = await res.json() as TemplateInfo
    if (!info.tar || !info.name) {
      throw new Error(`Invalid template info from ${registryURL}. name or tar fields are missing!`)
    }
    debug(`Fetched ${input} template info from ${registryURL} in ${Date.now() - start}ms`)
    return info
  })
}
