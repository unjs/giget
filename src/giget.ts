import { mkdir, rm } from "node:fs/promises";
import { existsSync, readdirSync } from "node:fs";
import { extract } from "tar";
import { resolve, dirname } from "pathe";
import { defu } from "defu";
import { cacheDirectory, download, debug } from "./_utils";
import { providers } from "./providers";
import { registryProvider } from "./registry";
import type { TemplateInfo, TemplateProvider } from "./types";

export interface DownloadTemplateOptions {
  provider?: string;
  force?: boolean;
  forceClean?: boolean;
  offline?: boolean;
  preferOffline?: boolean;
  providers?: Record<string, TemplateProvider>;
  dir?: string;
  registry?: false | string;
  cwd?: string;
  auth?: string;
}

const sourceProtoRe = /^([\w-.]+):/;

export type DownloadTemplateResult = Omit<TemplateInfo, "dir" | "source"> & {
  dir: string;
  source: string;
};

export async function downloadTemplate(
  input: string,
  options: DownloadTemplateOptions = {}
): Promise<DownloadTemplateResult> {
  options = defu(
    {
      registry: process.env.GIGET_REGISTRY,
      auth: process.env.GIGET_AUTH,
    },
    options
  );

  const registry =
    options.registry !== false
      ? registryProvider(options.registry, { auth: options.auth })
      : undefined;
  let providerName: string =
    options.provider || (registryProvider ? "registry" : "github");
  let source: string = input;
  const sourceProvierMatch = input.match(sourceProtoRe);
  if (sourceProvierMatch) {
    providerName = sourceProvierMatch[1];
    source = input.slice(sourceProvierMatch[0].length);
  }

  const provider =
    options.providers?.[providerName] || providers[providerName] || registry;
  if (!provider) {
    throw new Error(`Unsupported provider: ${providerName}`);
  }
  const template = await Promise.resolve()
    .then(() => provider(source, { auth: options.auth }))
    .catch((error) => {
      throw new Error(
        `Failed to download template from ${providerName}: ${error.message}`
      );
    });

  // Sanitize name and defaultDir
  template.name = (template.name || "template").replace(/[^\da-z-]/gi, "-");
  template.defaultDir = (template.defaultDir || template.name).replace(
    /[^\da-z-]/gi,
    "-"
  );

  const cwd = resolve(options.cwd || ".");
  const extractPath = resolve(cwd, options.dir || template.defaultDir);
  if (options.forceClean) {
    await rm(extractPath, { recursive: true, force: true });
  }
  if (
    !options.force &&
    existsSync(extractPath) &&
    readdirSync(extractPath).length > 0
  ) {
    throw new Error(`Destination ${extractPath} already exists.`);
  }
  await mkdir(extractPath, { recursive: true });

  const temporaryDirectory = resolve(
    cacheDirectory(),
    options.provider,
    template.name
  );
  const tarPath = resolve(
    temporaryDirectory,
    (template.version || template.name) + ".tar.gz"
  );

  if (options.preferOffline && existsSync(tarPath)) {
    options.offline = true;
  }
  if (!options.offline) {
    await mkdir(dirname(tarPath), { recursive: true });
    const s = Date.now();
    await download(template.tar, tarPath, {
      headers: {
        authorization: options.auth ? `Bearer ${options.auth}` : undefined,
        ...template.headers,
      },
    }).catch((error) => {
      if (!existsSync(tarPath)) {
        throw error;
      }
      // Accept network errors if we have a cached version
      debug("Download error. Using cached version:", error);
      options.offline = true;
    });
    debug(`Downloaded ${template.tar} to ${tarPath} in ${Date.now() - s}ms`);
  }

  if (!existsSync(tarPath)) {
    throw new Error(
      `Tarball not found: ${tarPath} (offline: ${options.offline})`
    );
  }

  const s = Date.now();
  const subdir = template.subdir?.replace(/^\//, "") || "";
  await extract({
    file: tarPath,
    cwd: extractPath,
    onentry(entry) {
      entry.path = entry.path.split("/").splice(1).join("/");
      if (subdir) {
        // eslint-disable-next-line unicorn/prefer-ternary
        if (entry.path.startsWith(subdir + "/")) {
          // Rewrite path
          entry.path = entry.path.slice(subdir.length);
        } else {
          // Skip
          entry.path = "";
        }
      }
    },
  });
  debug(`Extracted to ${extractPath} in ${Date.now() - s}ms`);

  return {
    ...template,
    source,
    dir: extractPath,
  };
}
