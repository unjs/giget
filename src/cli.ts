import { relative } from "node:path";
import { defineCommand, runMain } from "citty";
import { consola } from "consola";
import pkg from "../package.json" assert { type: "json" };
import { downloadTemplate } from "./giget";
import { startShell } from "./_utils";

const mainCommand = defineCommand({
  meta: {
    name: pkg.name,
    version: pkg.version,
    description: pkg.description,
  },
  args: {
    // TODO: Make it `-t` in the next major version
    template: {
      type: "positional",
      description:
        "Template name or a a URI describing provider, repository, subdir, and branch/ref",
    },
    dir: {
      type: "positional",
      description: "A relative or absolute path where to extract the template",
      required: false,
    },
    auth: {
      type: "string",
      description:
        "Custom Authorization token to use for downloading template. (Can be overriden with `GIGET_AUTH` environment variable)",
    },
    cwd: {
      type: "string",
      description:
        "Set current working directory to resolve dirs relative to it",
    },
    force: {
      type: "boolean",
      description: "Clone to existing directory even if exists",
    },
    forceClean: {
      type: "boolean",
      description:
        "Remove any existing directory or file recusively before cloning",
    },
    offline: {
      type: "boolean",
      description: "o not attempt to download and use cached version",
    },
    preferOffline: {
      type: "boolean",
      description: "Use cache if exists otherwise try to download",
    },
    shell: {
      type: "boolean",
      description: "Open a new shell with current working ",
    },
    install: {
      type: "boolean",
      description: "Install dependencies after cloning",
    },
    verbose: {
      type: "boolean",
      description: "Show verbose debugging info",
    },
  },
  run: async ({ args }) => {
    if (args.verbose) {
      process.env.DEBUG = process.env.DEBUG || "true";
    }

    const r = await downloadTemplate(args.template, {
      dir: args.dir,
      force: args.force,
      forceClean: args.forceClean,
      offline: args.offline,
      preferOffline: args.preferOffline,
      auth: args.auth,
      install: args.install,
    });

    const _from = r.name || r.url;
    const _to = relative(process.cwd(), r.dir) || "./";
    consola.log(`âœ¨ Successfully cloned \`${_from}\` to \`${_to}\`\n`);

    if (args.shell) {
      startShell(r.dir);
    }
  },
});

runMain(mainCommand);
