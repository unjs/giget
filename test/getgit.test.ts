import { existsSync } from "node:fs";
import { rm, mkdir, writeFile } from "node:fs/promises";
import { expect, it, describe, beforeAll } from "vitest";
import { resolve } from "pathe";
import { downloadTemplate } from "../src";

describe("downloadTemplate", () => {
  beforeAll(async () => {
    await rm(resolve(__dirname, ".tmp"), { recursive: true, force: true });
  });

  it("clone unjs/template", async () => {
    const destinationDirectory = resolve(__dirname, ".tmp/cloned");
    const { dir } = await downloadTemplate("gh:unjs/template", {
      dir: destinationDirectory,
      preferOffline: true,
    });
    expect(existsSync(resolve(dir, "package.json")));
  });

  it("do not clone to existing dir", async () => {
    const destinationDirectory = resolve(__dirname, ".tmp/existing");
    await mkdir(destinationDirectory).catch(() => {});
    await writeFile(resolve(destinationDirectory, "test.txt"), "test");
    await expect(
      downloadTemplate("gh:unjs/template", { dir: destinationDirectory }),
    ).rejects.toThrow("already exists");
  });
});
