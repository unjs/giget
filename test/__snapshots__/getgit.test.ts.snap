// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`downloadTemplate > clone unjs/template#e24616c using git provider (specific commit) 1`] = `
[
  "LICENSE",
  "README.md",
  "package.json",
  "renovate.json",
  "src",
  "src/index.ts",
  "test",
  "test/index.test.ts",
  "tsconfig.json",
]
`;
